src/trs/lexicon.rs

```diff
@@ -0,0 +1,818 @@
+use itertools::Itertools;
+use polytype::{Context as TypeContext, Type, TypeSchema, Variable as TypeVar};
+use rand::Rng;
+use std::collections::HashMap;
+use std::f64::NEG_INFINITY;
+use std::fmt;
+use std::iter;
+use std::sync::{Arc, RwLock};
+use term_rewriting::{
+    Atom, Context, Operator, Place, Rule, RuleContext, Signature, Term, Variable, TRS as UntypedTRS,
+};
+
+use super::{SampleError, TypeError, TRS};
+use utils::{logsumexp, weighted_permutation};
+use GP;
+
+#[derive(Debug, Clone)]
+/// Parameters for [genetic programming] with a [`Lexicon`].
+///
+/// [genetic programming]: struct.Lexicon.html#impl-GP
+/// [`Lexicon`]: struct.Lexicon.html
+pub struct GeneticParams {
+    /// The number of hypotheses crossover should generate.
+    pub n_crosses: usize,
+    /// The maximum number of non-progressing recursive steps sampling can make without failing.
+    pub max_sample_depth: usize,
+    /// The probability of adding (vs. deleting) a rule during mutation.
+    pub p_add: f64,
+    /// The probability of keeping a rule during crossover.
+    pub p_keep: f64,
+    /// If `true`, then the `TRS` should be deterministic.
+    pub deterministic: bool,
+    /// rule templates to use when sampling rules
+    pub templates: Vec<RuleContext>,
+    /// the weight to assign variables, constants, and non-constant operators, respectively.
+    pub atom_weights: (f64, f64, f64),
+}
+
+/// Manages the syntax of a term rewriting system.
+#[derive(Clone)]
+pub struct Lexicon(pub(crate) Arc<RwLock<Lex>>);
+impl Lexicon {
+    /// Construct a `Lexicon`.
+    pub fn new(
+        operators: Vec<(u32, Option<String>, TypeSchema)>,
+        background: Vec<Rule>,
+    ) -> Lexicon {
+        let mut signature = Signature::default();
+        let mut ops = Vec::with_capacity(operators.len());
+        for (id, name, tp) in operators {
+            signature.new_op(id, name);
+            ops.push(tp)
+        }
+        Lexicon(Arc::new(RwLock::new(Lex {
+            ops,
+            vars: Vec::new(),
+            signature,
+            background,
+        })))
+    }
+    /// Construct a `Lexicon`, including a set of background [`Variable`]s.
+    ///
+    /// [`Variable`]: ../../term_rewriting/struct.Variable.html
```

Comment:
Use docs.rs link, and say `term_rewriting::Variable` to make it clear that it's in another crate.